src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SSRP.cs

```diff
@@ -0,0 +1,151 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.Data.SqlClient.SNI
+{
+    internal class SSRP
+    {
+        private const char SemicolonSeparator = ';';
+        private const int SqlServerBrowserPort = 1434;
+
+        /// <summary>
+        /// Finds instance port number for given instance name.
+        /// </summary>
+        /// <param name="browserHostname">SQL Sever Browser hostname</param>
+        /// <param name="instanceName">instance name to find port number</param>
+        /// <returns>port number for given instance name</returns>
+        internal static int GetPortByInstanceName(string browserHostName, string instanceName)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(browserHostName), "browserHostName should not be null, empty, or whitespace");
+            Debug.Assert(!string.IsNullOrWhiteSpace(instanceName), "instanceName should not be null, empty, or whitespace");
+
+            byte[] instanceInfoRequest = CreateInstanceInfoRequest(instanceName);
+            byte[] responsePacket = SendUDPRequest(browserHostName, SqlServerBrowserPort, instanceInfoRequest);
+
+            const byte SvrResp = 0x05;
+            if (responsePacket == null || responsePacket.Length <= 3 || responsePacket[0] != SvrResp ||
+                BitConverter.ToUInt16(responsePacket, 1) != responsePacket.Length - 3)
+            {
+                throw new SocketException();
+            }
+
+            string serverMessage = Encoding.ASCII.GetString(responsePacket, 3, responsePacket.Length - 3);
+
+            string[] elements = serverMessage.Split(SemicolonSeparator);
+            int tcpIndex = Array.IndexOf(elements, "tcp");
+            if (tcpIndex < 0 || tcpIndex == elements.Length - 1)
+            {
+                throw new SocketException();
+            }
+
+            return ushort.Parse(elements[tcpIndex + 1]);
+        }
+
+        /// <summary>
+        /// Creates instance port lookup request (CLNT_UCAST_INST) for given instance name.
+        /// </summary>
+        /// <param name="instanceName">instance name to lookup port</param>
+        /// <returns>Byte array of instance port lookup request (CLNT_UCAST_INST)</returns>
+        private static byte[] CreateInstanceInfoRequest(string instanceName)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(instanceName), "instanceName should not be null, empty, or whitespace");
+
+            const byte ClntUcastInst = 0x04;
+            int byteCount = Encoding.ASCII.GetByteCount(instanceName);
+
+            byte[] requestPacket = new byte[byteCount + 1];
+            requestPacket[0] = ClntUcastInst;
+            Encoding.ASCII.GetBytes(instanceName, 0, instanceName.Length, requestPacket, 1);
```

Comment:
The spec for CLNT_UCAST_INST (https://msdn.microsoft.com/en-us/library/cc219746.aspx) says that the instance name is null-terminated. The code here does not include a null-terminator.